pylabrobot/pumps/calibration.py

```diff
@@ -0,0 +1,181 @@
+      return PumpCalibration.load_from_value(calibration, num_items)
+    if isinstance(calibration, str):
+      if calibration.endswith(".json"):
+        return PumpCalibration.load_from_json(calibration)
+      if calibration.endswith(".csv"):
+        return PumpCalibration.load_from_csv(calibration)
+      raise NotImplementedError("Calibration filetype not supported.")
+    return PumpCalibration()
+
+  @classmethod
+  def load_from_json(cls, file: str) -> PumpCalibration:
+    """ Load a calibration from a json file.
+
+    Args:
+      file: json file to load calibration from.
+
+    Returns:
+      PumpCalibration
+
+    Raises:
+      TypeError: if the calibration pulled from the json is not a dictionary or list.
+    """
+    with open(file, "rb") as f:
+      calibration = json.load(f)
+    if isinstance(calibration, dict):
+      calibration = {int(key): float(value) for key, value in calibration.items()}
+      return PumpCalibration.load_from_dict(calibration)
+    if isinstance(calibration, list):
+      return PumpCalibration(calibration)
+    raise TypeError(f"Calibration pulled from {file} is not a dictionary or list.")
+
+  @classmethod
+  def load_from_csv(cls, file: str) -> PumpCalibration:
+    """ Load a calibration from a csv file.
+
+    Args:
+      file: csv file to load calibration from.
+
+    Returns:
+      PumpCalibration
+    """
+
+    with open(file, encoding="utf-8", newline="") as f:
+      reader = csv.DictReader(f)
+      calibration = {int(row["0"]): float(row["1"]) for row in reader}
+    return PumpCalibration.load_from_dict(calibration)
+
+  @classmethod
+  def load_from_dict(cls, calibration: Dict[int, Union[int, float]]) -> PumpCalibration:
+    """ Load a calibration from a dictionary.
+
+    Args:
+      calibration: dictionary to load calibration from.
+
+    Returns:
+      PumpCalibration
+
+    Raises:
+      ValueError: if the calibration dictionary is not formatted correctly.
+    """
+
+    if not any(key == 0 for key in calibration.keys()):
+      if any(key == 1 for key in calibration.keys()):
+        calibration = {key - 1: value for key, value in calibration.items()}
+      else:
+        raise ValueError("Calibration dictionary keys must start at 0 or 1.")
+    if len(calibration) == 1:
+      calibration = {0: list(calibration.values())[0]}
+    if any(key < 0 for key in calibration.keys()) or any(key >= len(calibration) for key in
+                                                         calibration.keys()):
+      raise ValueError("Calibration dictionary keys must be non-negative and less than the "
+                       "length of the dictionary.")
+    if any(value <= 0 for value in calibration.values()):
+      raise ValueError("Calibration dictionary values must be positive.")
+    if sorted(calibration.keys()) != list(range(len(calibration))):
+      raise ValueError("Calibration dictionary keys must be a contiguous range of integers.")
```

Comment:
This is equivalent to the test above (l.134, `if any(key < 0...`) but a lot cleaner IMO. Do you want to keep this one and remove the other one?